src/Pact/Types/Crypto.hs

```diff
@@ -219,64 +167,59 @@ instance ConvertBS Ed25519.Signature where
   fromBS = Right . Ed25519.Sig
 #endif
 
-
-
---------- SCHEME ETH INSTANCES --------
-
-instance Scheme (SPPKScheme 'ETH) where
-  type PublicKey (SPPKScheme 'ETH) = ECDSA.PublicKey
-  type PrivateKey (SPPKScheme 'ETH) = ECDSA.PrivateKey
-  type Signature (SPPKScheme 'ETH) = ECDSA.Signature
-
-
-  _sign _ (Hash msg) pub priv = ECDSA.signETH (fromShort msg) pub priv
-  _valid _ (Hash msg) pub sig = ECDSA.validETH (fromShort msg) pub sig
-  _genKeyPair _ = ECDSA.genKeyPair
-  _getPublic _ = Just . ECDSA.getPublicKey
-  _formatPublicKey _ p = ECDSA.formatPublicKeyETH (toBS p)
-
-
-
-instance ConvertBS (ECDSA.PublicKey) where
-  toBS = ECDSA.exportPublic
-  fromBS s = maybeToEither ("Invalid ECDSA Public Key: " ++ show (toB16Text s))
-             (ECDSA.importPublic s)
-instance ConvertBS (ECDSA.PrivateKey) where
-  toBS = ECDSA.exportPrivate
-  fromBS s = maybeToEither ("Invalid ECDSA Private Key: " ++ show (toB16Text s))
-             (ECDSA.importPrivate s)
-instance ConvertBS (ECDSA.Signature) where
-  toBS = ECDSA.exportSignature
-  fromBS s = maybeToEither ("Invalid ECDSA Signature: " ++ show (toB16Text s))
-             (ECDSA.importSignature s)
-
-
-
+instance Scheme (SPPKScheme 'WebAuthn) where
+
+  _valid _ msg (PubBS pubBS) (SigBS sigBS) =
+    case runVerification of
+      Left _ -> False
+      Right () -> True
+    where
+      -- Verifying a WebAuthn signature requires that we know the payload
+      -- signed by the WebAuthn keys on the client device. This payload is
+      -- a combination of the Challenge (in our case, the PactHash of a
+      -- transaction), a JSON object of "client metadata", and data about
+      -- the authenticator hardware.
+      -- We require this data to be part of our WebAuthn signature so that
+      -- we can reconstitute the payload that was signed on the browser.
+      runVerification :: Either String ()
+      runVerification = do
+
+        -- Decode our WebAuthn signature object from a `UserSig` string.
+        WebAuthnSignature
+          { authenticatorData
+          , signature
+          , clientDataJSON } <- A.eitherDecode (BSL.fromStrict sigBS)
+
+        -- Decode the signer's public key.
+        publicKey <- first show $
+          Serialise.deserialiseOrFail @WA.CosePublicKey
+            (BSL.fromStrict pubBS)
```

Comment:
Should we restrict the algorithm here to just a few to avoid algorithm confusion attacks? I think @EnoF wanted the algorithm "ES256", which has COSE identifier `-7`.  https://portswigger.net/web-security/jwt/algorithm-confusion

Reply:
I've added some code that restricts to -7 or -8. (ECDSA SHA-256, and EdDSA respectively).

I informally tested the restriction by setting the requirement to -1 or -8 temporarily and observing that the happy path signature failed.

I'll hopefully be able to generate a test case that we can check in to the repo soon. But it requires spinning up a server that. implements webauthn to create new sample keys with a different signing algorithm.